core/src/main/java/org/elasticsearch/monitor/os/OsProbe.java

```diff
@@ -180,6 +188,294 @@ public short getSystemCpuPercent() {
+     *
+     * @param line the line to parse
+     * @param message a debug message to log if parsing fails
+     * @return the parsed value
+     */
+    private long parseSingleLineAsLong(final String line, final Supplier<String> message) {
+        if (line != null) {
+            try {
+                return Long.parseLong(line);
+            } catch (final NumberFormatException e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(message.get(), e);
+                }
+                // fallback
+            }
+        }
+
+        return -1;
+    }
+
+    // pattern for lines in /proc/self/cgroup
+    private static final Pattern CONTROL_GROUP_PATTERN = Pattern.compile("\\d+:([^:,]+(?:,[^:,]+)?):(/.*)");
+
+    /**
+     * A map of the control groups to which the Elasticsearch process
+     * belongs. Note that this is a map because the control groups can
+     * vary from subsystem to subsystem. Additionally, this map can not
+     * be cached because a running process can be reclassified.
+     *
+     * @return a map from subsystems to the control group for the
+     * Elasticsearch process.
+     */
+    private Map<String, String> getControlGroups() {
+        final List<String> lines = readProcSelfCgroup();
+        if (!lines.isEmpty()) {
+            final Map<String, String> controllerMap = new HashMap<>();
+            for (final String line : lines) {
+                final Matcher matcher = CONTROL_GROUP_PATTERN.matcher(line);
+                // note that Matcher#matches must be invoked as
+                // matching is lazy; this can not happen in an assert
+                // as assertions might not be enabled
+                if (!matcher.matches()) {
+                    assert false : line;
+                }
+                // at this point we have captured the subsystems and the
+                // control group
+                final String[] controllers = matcher.group(1).split(",");
+                for (final String controller : controllers) {
+                    controllerMap.put(controller, matcher.group(2));
+                }
+            }
+            return controllerMap;
+        }
+
+        return Collections.emptyMap();
+    }
+
+    /**
+     * The lines from {@code /proc/self/cgroup}. This file represents
+     * the control groups to which the Elasticsearch process belongs.
+     * Each line in this file represents a control group hierarchy of
+     * the form
+     * <p>
+     * {@code \d+:([^:,]+(?:,[^:,]+)?):(/.*)}
+     * <p>
+     * with the first field representing the hierarchy ID, the second
+     * field representing a comma-separated list of the subsystems
+     * bound to the hierarchy, and the last field representing the
+     * control group.
+     *
+     * @return the lines from {@code /proc/self/cgroup} or an empty
+     * list.
+     */
+    @SuppressForbidden(reason = "access /proc/self/cgroup")
```

Comment:
Can this be a follow-up? We have to push `Environment` all the way down from `Bootstrap` because we initialize the probes there before the security manager is installed.